### PR TITLE
[MRG] Annotation fix to allow Semicolons in Description

### PR DESCRIPTION
**Pull request by adam2392**
Thanks for contributing a pull request! Please make sure you have read the
[contribution guidelines](https://mne.tools/dev/install/contributing.html)
before submitting.

#### Reference issue
Fixes: #7287 

#### What does this implement/fix?
WIP, attempts to check if removing the ";" check breaks anything.

#### Additional information
See Issue for sample data.

### CODE REVIEW

**Comment by agramfort** [comment]
I suspect the issue comes from saving the file to fif. However I agree that we should not prevent ";" unless you want to save it to a .fif file.

**Comment by larsoner** [comment]
And even in that case we can probably sanitize with a substitution like `;` turns into `{SEMICOLON}`, and on read `{SEMICOLON}` gets turned into `;`. The only problem with this is the tiny possibility someone has `{SEMICOLON}` in their descriptions currently, but this probability is quite small and even if they has this, they can probably easily work around our substitution with a `my_str.replace(';', '{SEMICOLON}')` somewhere in their pipelines.

**Comment by agramfort** [comment]
agreed

>

**Comment by adam2392** [comment]
In `annotations.py:806`, 

```
            elif kind == FIFF.FIFF_COMMENT:
                description = tag.data.split(':')
                description = [d.replace(';', ':') for d in
                               description]
```

Is this already doing that?

**Comment by larsoner** [comment]
Maybe? I would need to look at the code more closely to know, but really what I would do first (before changing any MNE code) is write a little test (following test-driven development):
```
def test_annot_sanitizing(tmpdir):
    """Test description sanitizing."""
    annot = Annotations([0], [1], ['a;:b'])
    fname = str(tmpdir.join('custom-annot.fif'))
    annot.save(fname)
    annot_read = read_annotations(fname)
    _assert_annotations_equal(annot, annot_read)
```
On `master` this reproduces your failure, and it's what we want to fix, which is great.

Then with your commenting out of lines this happens:
```
larsoner@bunk:~/python/mne-python$ pytest mne/tests/test_annotations.py -k sanit
...
        annot_read = read_annotations(fname)
>       _assert_annotations_equal(annot, annot_read)
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 1 / 1 (100%)
E        x: array(['a;:b'], dtype='<U4')
E        y: array(['a::b'], dtype='<U4')
```
So it replaces `;` with `:` but does not restore it on read.

My suggestion above should get this to pass.

**Comment by adam2392** [comment]
So it turns out that when you save a `.fif` annotations, the ":" is the issue, and they are replaced by ";" because of this function in `write.py`. I assume that ValueError was added originally to check downstream when annotations are split into onset, value and description.

```
def write_name_list(fid, kind, data):
    """Write a colon-separated list of names.

    Parameters
    ----------
    data : list of strings
    """
    write_string(fid, kind, ':'.join(data))
```

The tests pass, but I'm wondering is it necessary to add some documentation to the docstrings regarding this behavior?

**Comment by agramfort** [comment]
I fear this will break reading files made with an older version of MNE

**Comment by larsoner** [comment]
> I fear this will break reading files made with an older version of MNE

I'm not sure what you mean -- the separator in the FIF file on disk will still be `:` -- it's just that other readers that didn't implement un-sanitizing would see `{COLON}` instead of `:` in their strings after reading from disk.

In any case, the changes in this PR are just to sanitize (and un-sanitize) descriptions `annotations` specifically, not globally in `write_name_list`. So the only ones being sanitized with `{COLON}` (and back) are annotations descriptions so the scope is much more localized than a global `write_name_list` change would be.

Basically before the `:` character could be used but `;` could not. This PR essentially would change it so that instead of `;` being unusable, the sequence of characters `{COLON}` cannot be used (or can, but will be changed to `:` on I/O round-trip).

**Comment by agramfort** [comment]
@larsoner if you think it can be merged fair enough.

File: mne/annotations.py [hunk 1]
```diff
@@ -46,8 +46,6 @@ def _check_o_d_s(onset, duration, description):
     if description.ndim != 1:
         raise ValueError('Description must be a one dimensional array, '
                          'got %d.' % (description.ndim,))
-    if any([';' in desc for desc in description]):
```
Comment (larsoner) [inline mne/annotations.py]: Instead of removing these lines I guess we should change them to `if any('{COLON}' in desc for desc in description)`, then ideally also add a tiny test:
```
with pytest.raises(ValueError, match='in descriptions not supported'):
    Annotations([0], [1], 'a{COLON}b')
```
or so

File: mne/annotations.py [hunk 2]
```diff
@@ -46,8 +46,9 @@ def _check_o_d_s(onset, duration, description):
...
-        raise ValueError('Semicolons in descriptions not supported.')
+    if any(['{COLON}' in desc for desc in description]):
+        raise ValueError('{COLON} character '
```
Comment (larsoner) [inline mne/annotations.py]:
```suggestion
        raise ValueError('The substring "{COLON}" '
```

**Comment by adam2392** [comment]
@larsoner cool sounds good! fixed it up. Thanks.

**Comment by larsoner** [comment]
Thanks @adam2392 !

**Comment by larsoner** [comment]
Ahh, forgot `changes/latest.inc` update, can you open a quick PR to add a line there in the `BUG` section about this fix?

**Comment by nbara** [comment]
> I fear this will break reading files made with an older version of MNE

This is essentially what happened here though ? I had `.fif` files with `:`, which got converted to `;` before this PR. 

Now loading these `.fif` all have `;` instead of `:` 

Am I missing something @larsoner ?

**Comment by larsoner** [comment]
Yes that might still be the case :(

Would a workaround like this work for you? It should be backward compatible / work on 0.19 and 0.20:
```
for di, desc in enumerate(annotations.description):
    annotations.description[di] = desc.replace(';', ':')
```
In principle we could add a `replace_semicolon=None` kwarg which defaults to True in 0.20, False in 0.21, and disappears in 0.22, and only warns if there are actually semicolons in the descriptions. I guess this would need to be added to `read_raw_fif` and `read_annotations`. WDYT @nbara ?

**Comment by nbara** [comment]
Yeah I guess I'll be using this workaround, until I convert my `.fif` files to be 0.20 compatible.

The `replace_semicolon` kwarg seems a bit overkill, especially if it's to be removed in future versions.
